File: Codigos estudiantes por lenguaje/Java/Julian Silva/Strings/AnagramPattern.java
public class ejercicios {
	static final int MAX = 256;
// Esta función retorna true si los contenidos
// de arr1[] y arr2[] son iguales, de otra forma es falso

	static boolean compare(char vec1[], char vec2[]) {
		for (int i = 0; i < MAX; i++) {
			if (vec1[i] != vec2[i]) {
				return false;
			}
		}
		return true;
	}
// Esta función busca todas las permutaciones de
// pat[] en txt[]

	static void search(String busqueda, String str) {
		int x = busqueda.length();
		int y = str.length();
// countP[]:	Almacena el conteo de todos
// los caracteres del patrón
// countTW[]: Almacena el conteo de los caracteres
// la ventana de texto
		char[] conteopalabras = new char[MAX];
		char[] countTW = new char[MAX];
		for (int i = 0; i < x; i++) {
			(conteopalabras[busqueda.charAt(i)])++;
			(countTW[str.charAt(i)])++;
		}
//Atravesar atravez de los caracteres restantes de patrón
		for (int i = x; i < y; i++) { // Compara conteos de la ventana actual
// de texto con los conteos de pattern[]

			if (compare(conteopalabras, countTW)) {
				System.out.println("Encontrado en indice " + (i - x));
			}
// Agreca el actual caracter a la ventana actual 
			(countTW[str.charAt(i)])++;
// Remueve el primer caracter de la anterior ventana 
			countTW[str.charAt(i - x)]--;
		}
// Revisa por la ultima ventana en el texto 
		if (compare(conteopalabras, countTW)) {
		System.out.println("Encontrado en indice " + (y - x));
	}

}

	public static void main(String args[]) {
		String str = "BACDGABCDA";
		String busqueda = "ABCD";
		search(busqueda, str);
	}
}

File: Codigos estudiantes por lenguaje/C++/Julian_silva/Strings/AnagramPattern.cpp
using namespace std;
#include<bits/stdc++.h>
#include<cstdlib>
#define FAST ios_base::sync_with_stdio(false);cin.tie(NULL);cout.tie(0);
int MAX=256;
bool compare(char arr1[], char arr2[]) {
    for (int i = 0; i < MAX; i++) {
        if (arr1[i] != arr2[i]) {
            return false;
        }
    }
    return true;
}
void anagramsSearch(string pat,string txt){
    int M=pat.size();
    int N=txt.size();
    char countP[MAX];
    char countTW[MAX];
    memset(countP,0,sizeof countP);
    memset(countTW,0,sizeof countTW);
    for(int i=0;i<M;i++){
        (countP[pat[i]])++;
        (countTW[txt[i]])++;
    }
    for (int i = M; i <N; i++) {
            if (compare(countP, countTW)) {
                cout<<"Encontrado en indice "<<(i - M)<<" "<<txt.substr(i-M,M)<<endl;
            }
            countTW[txt[i]]++;
            countTW[txt[i - M]]--;
        }
        if (compare(countP, countTW)) {

           cout<<"Encontrado en indice "<<(N - M)<<" "<<txt.substr(N-M,N)<<endl;
    }
}
int main() {
    string txt ="BACDGABCDA ";
    string pat="ABCD";
    anagramsSearch(pat,txt);
}
